### DISCUSSION TITLE
Compartments and ODEs

### DISCUSSION THREAD

**Discussion by bjornhald** [sys-bio/tellurium#590] [Q&A]
Hi,
Thanks for a really nice and flexible environment for sys-bio modelling & simulations.

I want to report a small "bug" that caused some headache re compartments. I'm unsure whether it relates to tellurium or roadrunner. Nevertheless,  if I write up the same reactions for a standard 2-compartment PK model and load in models with without declaration of species in compartments, I get the same return when I ask for return of ODEs, but I also get different simulations! 
```

s_pk = """
species Aabs = 0.0;          # drug SC
species Ac = 0.0;              # drug central
species Ap = 0.0;              # drug peripheral

compartment Vc = 1.34
compartment Vp = 0.35
compartment SC = 1

const ka = 1.23
const Q = 0.15     # L/h
const Cl = 0.62
const F = 1.0

Aabs -> Ac;  ka * Aabs                 # [nmol/h] absorption 
Ac -> Ap    ;   Q * (Ac - Ap)          # [nmol/h] 
Ac ->          ;   Cl * Ac          
"""

pk_comp = """
species Ac in Vc;
species Ap in Vp; 
species Aabs in SC; 
"""

dose = 5
r = te.loada(s_pk + pk_comp)
r.Aabs += r.F * dose  # SC dosis
y = r.simulate(0, 15, 151)         

r2 = te.loada(s_pk )
r2.Aabs += r2.F * dose  # SC dosis
y2 = r2.simulate(0, 15, 151)       
```
If I then ask for the ODEs
```
>>> te.getODEsFromModel(r))
>>> te.getODEsFromModel(r2))
```

both yields 

```
v_J0 = ka*Aabs*SC
v_J1 = k12*Ac*Vc-k21*Ap*Vp
v_J2 = Ke*Ac*Vc

dAabs/dt = -v_J0
dAc/dt = v_J0 - v_J1 - v_J2
dAp/dt = v_J1

```
A plot of the two simulations, however, shows the difference. This is troubling. I can deduce that declaration of species in compartments makes tellurium divide the ODE of the relevant species with the volume of the compartment. But I cannot find info in the documentation. 
![image](https://github.com/user-attachments/assets/3ac3b8b5-fbe3-4549-9d80-9ab9a2cc9daf)

Is there a way of handling dosing without having to write up reactions as amount/time?

Thanks in advance

**Comment by luciansmith** [comment]
Ack, sorry for missing this earlier.

There is a bug in the 'getODEsFromModel' function that didn't include the compartments in the equations (https://github.com/sys-bio/roadrunner/issues/1256) This is fixed internally, but not released yet, though that should be soon.

This didn't affect anything else; the only bug was in that function (the ODEs were generated from scratch there but not used anywhere else).

To answer your final question, however:  "Is there a way of handling dosing without having to write up reactions as amount/time?":  The answer is no!  Reactions in SBML are defined as amount/time (or to be even more technical, 'extent/time').  This definition allows compartments to be different sizes and it allows compartment volumes to change over time while still being mathematically accurate.

If you *must* write something as concentration/time, then you have to use 'rate rules' instead.  In Antimony, you'd write this as something like:

```
Aabs' = k1*Aabs
```

I wouldn't recommend it, as you lose the other types of analysis available for your model when you use reactions, and if someone does end up wanting to use changing compartment sizes, the math might end up being wrong.  But it's possible!

You do also have to make sure that Aabs is either a parameter, or a 'concentration-type' species--in SBML, those are set 'hasOnlySubstanceUnits=false'.  But species are that type by default in Antimony, and you have to add a new declaration to change it:

```
substanceOnly species A2 in C
```
